From 58602e300aa1190d80b37e45bb98ab7c5bc940b8 Mon Sep 17 00:00:00 2001
From: William Douglas <william.douglas@intel.com>
Date: Fri, 27 Jan 2017 23:36:21 +0000
Subject: [PATCH] Add ability to pass alternate swupd cert

Swupd now enforces valid certificates when installing content. Because
of this, when installing content that was not generated with the
certificate of the currently running system swupd will fail. To work
around this swupd added the '-C' option to pass an alternate
certificate. Support this use case in ister by adding the --cert-file
option which will take a path to the alternate certificate file.
---
 ister.py      |  5 +++++
 ister_test.py | 39 +++++++++++++++++++++++++++++++++++++++
 2 files changed, 44 insertions(+)

diff --git a/ister.py b/ister.py
index 5f4ff72..0d4c99a 100755
--- a/ister.py
+++ b/ister.py
@@ -400,6 +400,8 @@ def copy_os(args, template, target_dir):
     swupd_command += " --versionurl={0}".format(args.versionurl)
     swupd_command += " --format={0}".format(args.format)
     swupd_command += " --statedir={0}".format(args.statedir)
+    if args.cert_file:
+        swupd_command += " -C {0}".format(args.cert_file)
 
     if template["DestinationType"] == "physical":
         os.makedirs("/var/lib/swupd", exist_ok=True)
@@ -1280,6 +1282,9 @@ def handle_options():
     parser.add_argument("-c", "--config-file", action="store",
                         default=None,
                         help="Path to configuration file to use")
+    parser.add_argument("-s", "--cert-file", action="store",
+                        default=None,
+                        help="Path to certificate file used by swupd")
     parser.add_argument("-t", "--template-file", action="store",
                         default=None,
                         help="Path to template file to use")
diff --git a/ister_test.py b/ister_test.py
index b780037..99b01ac 100644
--- a/ister_test.py
+++ b/ister_test.py
@@ -1232,6 +1232,7 @@ def copy_os_good():
     args.versionurl = "vtest"
     args.format = "formattest"
     args.statedir = "/statetest"
+    args.cert_file = None
     swupd_cmd = "swupd verify --install --path=/ --manifest=0 "              \
                 "--contenturl=ctest --versionurl=vtest --format=formattest " \
                 "--statedir=/statetest"
@@ -1245,6 +1246,39 @@ def copy_os_good():
 
 
 @run_command_wrapper
+def copy_os_cert_good():
+    """Check installer command with cert present"""
+    backup_add_bundles = ister.add_bundles
+    ister.add_bundles = lambda x, y: None
+    backup_which = shutil.which
+    shutil.which = lambda x: False
+
+    def args():
+        """args empty object"""
+        None
+    args.contenturl = "ctest"
+    args.versionurl = "vtest"
+    args.format = "formattest"
+    args.statedir = "/statetest"
+    args.cert_file = "/certtest"
+    swupd_cmd = "swupd verify --install --path=/ --manifest=0 "              \
+                "--contenturl=ctest --versionurl=vtest --format=formattest " \
+                "--statedir=/statetest -C /certtest"
+    commands = [swupd_cmd,
+                "https://to.clearlinux.org",
+                True]
+    template = {
+        "Version": 0,
+        "DestinationType": "",
+        "HTTPSProxy": "https://to.clearlinux.org"
+    }
+    ister.copy_os(args, template, "/")
+    ister.add_bundles = backup_add_bundles
+    shutil.which = backup_which
+    commands_compare_helper(commands)
+
+
+@run_command_wrapper
 def copy_os_proxy_good():
     """Check installer command with proxy present"""
     backup_add_bundles = ister.add_bundles
@@ -1259,6 +1293,7 @@ def copy_os_proxy_good():
     args.versionurl = "vtest"
     args.format = "formattest"
     args.statedir = "/statetest"
+    args.cert_file = None
     swupd_cmd = "swupd verify --install --path=/ --manifest=0 "              \
                 "--contenturl=ctest --versionurl=vtest --format=formattest " \
                 "--statedir=/statetest"
@@ -1294,6 +1329,7 @@ def copy_os_format_good():
     args.versionurl = "vtest"
     args.format = None
     args.statedir = "/statetest"
+    args.cert_file = None
     swupd_cmd = "swupd verify --install --path=/ --manifest=0 "        \
                 "--contenturl=ctest --versionurl=vtest --format=test " \
                 "--statedir=/statetest"
@@ -1323,6 +1359,7 @@ def copy_os_which_good():
     args.versionurl = "vtest"
     args.format = "formattest"
     args.statedir = "/statetest"
+    args.cert_file = None
     swupd_cmd = "swupd verify --install --path=/ --manifest=0 "              \
                 "--contenturl=ctest --versionurl=vtest --format=formattest " \
                 "--statedir=/statetest"
@@ -1353,6 +1390,7 @@ def copy_os_physical_good():
     args.versionurl = "vtest"
     args.format = "formattest"
     args.statedir = "/statetest"
+    args.cert_file = None
     swupd_cmd = "swupd verify --install --path=/ --manifest=0 "              \
                 "--contenturl=ctest --versionurl=vtest --format=formattest " \
                 "--statedir=/statetest"
@@ -4701,6 +4739,7 @@ if __name__ == '__main__':
         get_current_format_good,
         set_hostname_good,
         copy_os_good,
+        copy_os_cert_good,
         copy_os_proxy_good,
         copy_os_format_good,
         copy_os_which_good,
-- 
2.11.0

